packages/api-providers/src/web3/vanchor-withdraw.ts

```diff
@@ -450,4 +401,68 @@ export class Web3VAnchorWithdraw extends VAnchorWithdraw<WebbWeb3Provider> {
       outputNotes: changeNotes,
     };
   }
+
+  private async fetchNoteLeaves(
+    note: string,
+    leavesMap: Record<string, Uint8Array[]>,
+    destVAnchor: VAnchorContract,
+    treeHeight: number
+  ): Promise<{ leafIndex: number; utxo: Utxo; amount: BigNumber }> {
+    const abortSignal = this.cancelToken.abortSignal;
+    const parsedNote = (await Note.deserialize(note)).note;
+    const amount = BigNumber.from(parsedNote.amount);
+
+    // fetch leaves if we don't have them
+    if (leavesMap[parsedNote.sourceChainId] === undefined) {
+      // Set up a provider for the source chain
+      const sourceAddress = parsedNote.sourceIdentifyingData;
+      const sourceChainConfig = this.config.chains[Number(parsedNote.sourceChainId)];
+      const sourceHttpProvider = Web3Provider.fromUri(sourceChainConfig.url);
+      const sourceEthers = sourceHttpProvider.intoEthersProvider();
+      const sourceVAnchor = this.inner.getVariableAnchorByAddressAndProvider(sourceAddress, sourceEthers);
+      const leafStorage = await bridgeStorageFactory(Number(parsedNote.sourceChainId));
+      const leaves = await this.inner.getVariableAnchorLeaves(sourceVAnchor, leafStorage, abortSignal);
+
+      leavesMap[parsedNote.sourceChainId] = leaves.map((leaf) => {
+        return hexToU8a(leaf);
+      });
+    }
```

Comment:
The mutation of the `leavesMap` object is being done with many async calls this can make undermind behavior, and some values could be overwritten.
it will be better to prepare the map with default values before mutation, and while mutating consider adding to the value
 **EX :**
This call 
```typescript
leavesMap[parsedNote.sourceChainId] = leaves.map((leaf) => {
        return hexToU8a(leaf);
      });
```
could be 
```typescript
leavesMap[parsedNote.sourceChainId] =[ ...leavesMap[parsedNote.sourceChainId],leaves.map((leaf) => {
        return hexToU8a(leaf);
      })];
```

Besides the fact that you may fetch the same leaves twice

Reply:
@nepoche  wanna know you thoughts as well

Reply:
I agree that the check:
```
    // fetch leaves if we don't have them
    if (leavesMap[parsedNote.sourceChainId] === undefined) {
```
will not prevent multiple calls for fetching leaves of the same chain, and the leavesMap[parsedNote.sourceChainId] will be overwritten with leaf data multiple times.  However, the data should always be all leaves 1x.  With the 2nd option you listed, the data would be all the leaves potentially 1-15x.

Reply:
I think the proposal from @AhmedKorim will lead to more bugs, it could end up appending lots of duplicate leaves. I do think that there's room for improvement on ensuring we only fetch leaves for each unique chain once.

We could go through all notes and get the unique set of chains they're originating/targeting **and then** begin querying each one. This would ensure we don't double-start queries from the same chain before either finishes.

Reply:
Can we try and put each chain query into a WebWorker? Do you think that would help parallelize and not block the main thread @AtelyPham ?

Reply:
Yeah, I agree that we can prepare a set of unique source chains and then fetch leaves from the prepared set. About the `WebWorker`, I think we don't really need the WebWorker, because the main thread still needs the result of the parallel computations to continue.